### DISCUSSION TITLE
How to test in GraphQL Spring Boot

### DISCUSSION THREAD

**Discussion by alioth865** [graphql-java-kickstart/graphql-spring-boot#267] [Q&A]
Hello, im new in GraphQL world, I'm doing my first API with this technology and I have one doubt. It's possible with JUnit do testing for my GraphQLResolver, GraphQLMutationResolver or GraphQLQueryResolver? Or simply don't have sence doing? If is possible doing, please can you share me someone class than example? 
Sorry for my bad english! 
Best regards

**Comment by oliemansm** [comment]
As mentioned in the documentation there's a test starter available which exposes the `@GraphQLTest` annotation. Take a look at https://github.com/graphql-java-kickstart/graphql-spring-boot/blob/master/example-graphql-tools/src/test/java/com/graphql/sample/boot/GraphQLToolsSampleApplicationTest.java for an example how to use it.

**Reply by Cat-Lord** [reply]
For anyone trying to access the test in the link, [here is an updated version](https://github.com/graphql-java-kickstart/samples/blob/master/tools-spring-boot/src/test/java/com/graphql/sample/boot/GraphQLToolsSampleApplicationTest.java).

You can find samples in the [a separate repository](https://github.com/graphql-java-kickstart/samples).

**Reply by yusuffurkanyalcin** [reply]
Hello , 

I just write like this code but it is not working on my code


@ExtendWith(SpringExtension.class)
@GraphQLTest
public class PlayerMutationIntTest {

    private static final String GRAPHQL_MUTATION_REQUEST_PATH = "graphql/request/create-player.graphql";

    @Autowired
    private GraphQLTestTemplate graphQLTestTemplate;


    @Test
    public void create() throws IOException{
            GraphQLResponse response = graphQLTestTemplate.postForResource(GRAPHQL_MUTATION_REQUEST_PATH);
            System.out.println(response.getRawResponse().getBody());
    }

}

**Reply by yusuffurkanyalcin** [reply]
My response is : 

<!doctype html><html lang="en"><head><title>HTTP Status 404 – Not Found</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 404 – Not Found</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Description</b> The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.</p><hr class="line" /><h3>Apache Tomcat/9.0.56</h3></body></html>

**Comment by anrolmar** [comment]
Hi. I used the example, but I retrieved a 404 error. If I look around the code, the error is this piece of code: 
`ResponseEntity<String> response = this.restTemplate.exchange(this.graphqlMapping, HttpMethod.POST, request, String.class, new Object[0]);`

`this.graphqlMapping` has the value /graphql.

Is this correct?

**Comment by FranBisquerra** [comment]
Hi! I'm trying to use the test library as I found it pretty easy and useful and I'm facing the same 404 error.. In my case I found that is related with the

`@RunWith(SpringRunner::class) @GraphQLTest`

annotations, if I use the 

`@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)` 

it works perfectly, I can't seem to find the reason as `@GraphQLTest` is supposed to configure the environment the same way. 

Anyone that knows what the problem might be?

Thx!

**Reply by arjun-oodle** [reply]
When you add the above code, ./gradlew test hangs. Unlike the Injteli J IDE where it runs perfectly.

**Comment by oliemansm** [comment]
[This test class](https://github.com/graphql-java-kickstart/graphql-spring-boot/blob/4c7cdbaa8b576134bad0370add9241207e267904/example-graphql-tools/src/test/java/com/graphql/sample/boot/GraphQLToolsSampleApplicationTest.java) uses `@GraphQLTest` too without specifying RANDOM_PORT specifically either and is working fine.

You shouldn't have to add this:
```java
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
since that's already set on the `@GraphQLTest` annotation.

Maybe it's a Kotlin config issue?

**Comment by FranBisquerra** [comment]
Hi! I've been trying for a while to find the problem but didn't find anything.. Reading the annotation doc I realized that It was not loading the full application context, and I need it anyway, so I ended up using the `@SpringBootTest` annotation. 

Thank you for the support!

**Comment by imamika** [comment]
Hello! I am working on building GraphQL Webflux api using graphql-kickstart-spring-boot-starter-webflux. I tried finding resources on how i could test this. Any pointers in this regard are helpful. Thanks!

**Comment by imamika** [comment]
I think i found what i am looking for. 

https://github.com/graphql-java-kickstart/graphql-spring-webclient/blob/master/graphql-webclient/src/test/java/graphql/kickstart/spring/webclient/boot/GraphQLWebClientTest.java

**Comment by mustafa-qamaruddin** [comment]
I am trying the following test:
```
@GraphQLTest
public class UserTest {
  @Autowired
  private GraphQLTestTemplate graphQLTestTemplate;

  @Test
  public void createUser() throws IOException {
    GraphQLResponse response = graphQLTestTemplate.postForResource("graphql/register.graphql");
    assertNotNull(response);
    assertTrue(response.isOk());
    assertEquals("1", response.get("$.data.post.id"));
  }
}
```
and I get the following response:
```
<404,<!doctype html><html lang="en"><head><title>HTTP Status 404 – Not Found</title><style type="text/css">body {font-family:Tahoma,Arial,sans-serif;} h1, h2, h3, b {color:white;background-color:#525D76;} h1 {font-size:22px;} h2 {font-size:16px;} h3 {font-size:14px;} p {font-size:12px;} a {color:black;} .line {height:1px;background-color:#525D76;border:none;}</style></head><body><h1>HTTP Status 404 – Not Found</h1><hr class="line" /><p><b>Type</b> Status Report</p><p><b>Message</b> The requested resource [&#47;api&#47;graphql] is not available</p><p><b>Description</b> The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.</p><hr class="line" /><h3>Apache Tomcat/9.0.37</h3></body></html>,[Content-Type:"text/html;charset=utf-8", Content-Language:"en", Content-Length:"766", Date:"Sun, 08 Nov 2020 00:18:11 GMT", Keep-Alive:"timeout=60", Connection:"keep-alive"]>
```

Any help is appreciated ...

**Reply by rizvisar** [reply]
I am getting the same response. @mustafa-qamaruddin were you able to get past this ??

**Comment by EriRan** [comment]
@mustafa-qamaruddin That test class is correct. There is something wrong with the servlet, possibly due to the graphql not being available in *URI*/graphql

**Comment by mustafa-qamaruddin** [comment]
Thank you, I have my /graphql pointing to /api/graphql, shouldn't the test detect it automaticaly, and from debugging I think it hits /api/graphql
```
graphql:
  tools:
    schema-location-pattern: "**/*.graphqls"
    introspection-enabled: true
  servlet:
    mapping: /api/graphql
    enabled: true
```
Also, I @GraphQLTest should be initializing the servlet?

**Comment by chenbocheng** [comment]
I had the same issue. However, once I just put my test class into the same package, then it worked.

**Reply by youaresodalgona** [reply]
Can you explain more?

**Reply by lloyd137** [reply]
@cr-lgl Did you maybe find solution yet? :)

**Reply by youaresodalgona** [reply]
@lloyd137 Don't use `@GraphqlTest` annotation, use `@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)`.

look [the comment](https://github.com/graphql-java-kickstart/graphql-spring-boot/discussions/267#discussioncomment-226020)

**Reply by lloyd137** [reply]
@cr-lgl Thanks for your response! I've tried that before however, if I use the @'s you mentioned I this error:
`java.lang.IllegalStateException: Failed to load ApplicationContext`

`Also this is printed in the stack:
```
'***************************
APPLICATION FAILED TO START
***************************

Description:

The bean 'schemaStringProvider', defined in class path resource [graphql/kickstart/tools/boot/GraphQLJavaToolsAutoConfiguration.class], could not be registered. A bean with that name has already been defined in class path resource [com/oembedler/moon/graphql/boot/GraphQLJavaToolsAutoConfiguration.class] and overriding is disabled.

Action:

Consider renaming one of the beans or enabling overriding by setting spring.main.allow-bean-definition-overriding=true'
`
```

I honestly do not know what is meant by the action, as in where i can change that and what for effect it'll have.

Do you know what this means?

**Reply by youaresodalgona** [reply]
Well, it’s the first time I’ve seen...

`com/oembedler/moon/graphql/boot/GraphQLJavaToolsAutoConfiguration.class` Which library is this file in?

Maybe... It  seems to be a problem caused by trying to register two beans with the same name.

**Reply by oliemansm** [reply]
Something is pulling in a very old version of one of the libraries that still used the `com.oembedler.moon` package. Older versions of `com.zhokhov.graphql:graphql-datetime-spring-boot-starter` can do that for example. Make sure that you only have the compatible versions. When using the spring-boot-starter you don't need to define any other kickstart dependencies, since it pulls in the correct and compatible versions for you.

**Reply by lloyd137** [reply]
@oliemansm @cr-lgl Okay so what I think you mean is that I'm using a not compatible library for my project. These are the dependencies I'm using, do you maybe see which ones aren't 'correct'?

        <dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-spring-boot-starter</artifactId>
            <version>5.0.2</version>
        </dependency>
        <dependency>
            <groupId>com.graphql-java</groupId>
            <artifactId>graphql-java-tools</artifactId>
            <version>5.2.4</version>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>30.1-jre</version>
        </dependency>
        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
        </dependency>

For testing:

        <dependency>
            <groupId>com.graphql-java-kickstart</groupId>
            <artifactId>graphql-spring-boot-starter-test</artifactId>
            <version>7.0.1</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.vintage</groupId>
            <artifactId>junit-vintage-engine</artifactId>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.hamcrest</groupId>
                    <artifactId>hamcrest-core</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
    </dependencies>

**Reply by oliemansm** [reply]
Yes, you're using incompatible versions. Remove graphql-java-tools (the second one) entirely and change the group and version of graphql-spring-boot-starter to the same values as graphql-spring-boot-starter-test. So group "com.graphql-java-kickstart" and version 7.0.1.

You're main GraphQL package is depending on a very old and outdated version.

**Reply by lloyd137** [reply]
@oliemansm It works now! I appreciate your help a lot, thankyou!! :)

and @cr-lgl thankyou too!

**Reply by arjun-oodle** [reply]
`./gradlew test` hangs when I do this. But runs perfect through the intelij IDE. Unable to deploy as jenkins fails because of `./gradlew test`

**Comment by Predarion** [comment]
I had the same issue and solved it with:
In my pom.xml:
```
<!-- Spring Webflux server -->
<dependency>
	<groupId>org.springframework.boot</groupId>
	<artifactId>spring-boot-starter-webflux</artifactId>
</dependency>

<!-- GraphQL --> 
<dependency>
	<groupId>com.graphql-java</groupId>
	<artifactId>graphql-java-spring-boot-starter-webflux</artifactId>
	<version>2.0</version>
</dependency>
<dependency>
	<groupId>com.graphql-java-kickstart</groupId>
	<artifactId>graphql-java-tools</artifactId>
	<version>11.0.0</version>
</dependency>

<!-- Testing -->
<dependency>
	<groupId>com.graphql-java-kickstart</groupId>
	<artifactId>graphql-spring-boot-starter-test</artifactId>
	<version>11.0.0</version>
	<scope>test</scope>
</dependency>
```

Then, I can test with
```
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class QueryResolverIT {
    @Autowired
    private GraphQLTestTemplate template;

    @Test
    void testAffiliation() throws IOException {
        GraphQLResponse response = template.postForResource("graphql/query/affiliation.graphql");
        Assertions.assertThat(response).isNotNull();
        Assertions.assertThat(response.isOk()).isTrue();
    }
}
```

`@GraphqlTest` didn't work for me.

**Comment by arjun-oodle** [comment]
Adding this works for me `@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)`.
But the test fail on jenkins with weird issue.

```
13:29:58  > Task :test
13:29:58  
13:29:58  ExternalQueryResolverTest > testValidAddressEndpoint() FAILED
13:29:58      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:58          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:58              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:29:58  
13:29:58  ExternalQueryResolverTest > testValidAddressesEndpoint() FAILED
13:29:58      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:58          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:58              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:29:58  
13:29:58  ExternalQueryResolverTest > testValidAddressEndpointUnableToFetchAddressException() FAILED
13:29:58      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:58          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:58              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:29:58  
13:29:58  ExternalQueryResolverTest > testValidAddressesEndpointUnableToFetchAddressException() FAILED
13:29:58      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:58          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:58              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:29:58  
13:29:58  ExternalQueryResolverTest > testValidAddressEndpointInvalidPostcode() FAILED
13:29:58      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:58          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:58              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:29:59  
13:29:59  ExternalQueryResolverTest > testValidAddressesEndpointInvalidPostcode() FAILED
13:29:59      java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:132
13:29:59          Caused by: org.springframework.beans.factory.BeanDefinitionStoreException at ConfigurationClassParser.java:189
13:29:59              Caused by: java.io.FileNotFoundException at ClassPathResource.java:180
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  2021-04-16 08:00:47.778  INFO 7147 --- [extShutdownHook] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'applicationTaskExecutor'
13:30:49  
13:30:49  260 tests completed, 6 failed
```

The test that passed are for Rest API. GraphQL tests are failing.

Can anyone help?

**Comment by ermadmi78** [comment]
Hi, you can find GraphQL Spring Boot server example here:
https://github.com/ermadmi78/kobby-gradle-example

Tests example is placed here:
https://github.com/ermadmi78/kobby-gradle-example/blob/main/cinema-server/src/test/kotlin/io/github/ermadmi78/kobby/cinema/server/CinemaServerTest.kt

To run tests from IntelliJ IDEA install Kotest plugin:
https://plugins.jetbrains.com/plugin/14080-kotest
